README.md
# Managed Identity and Key Vault with ASP.NET Core

> Build an ASP.NET Core Web API using Managed Identity, Key Vault, and Cosmos DB that is designed to be deployed to Azure App Service or Azure Kubernetes Service (AKS) as a Docker container

![License](https://img.shields.io/badge/license-MIT-green.svg)
![Docker Image Build](https://github.com/retaildevcrews/helium-csharp/workflows/Docker%20Image%20Build/badge.svg)

This is an ASP.NET Core Web API reference application designed to "fork and code" with the following features:

- Securely build, deploy and run an Azure App Service (Web App for Containers) application
- Securely build, deploy and run an Azure Kubernetes Service (AKS) application
- Use Managed Identity to securely access resources
- Securely store secrets in Key Vault
- Securely build and deploy the Docker container to Azure Container Registry (ACR) or Docker Hub
- Connect to and query Cosmos DB
- Automatically send telemetry and logs to Azure Monitor

> GitHub Codespaces is the easiest way to evaluate helium as all of the prerequisites are automatically installed
>
> Follow the setup steps in the [Helium readme](https://github.com/retaildevcrews/helium) to setup Codespaces

## Prerequisites

- Bash shell (tested on GitHub Codespaces, Mac, Ubuntu, Windows with WSL2)
  - Will not work with WSL1 or Cloud Shell
- .NET Core SDK 3.1 ([download](https://dotnet.microsoft.com/download))
- Docker CLI ([download](https://docs.docker.com/install/))
- Azure CLI ([download](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest))
- Visual Studio Code (optional) ([download](https://code.visualstudio.com/download))

## Setup

- Initial setup instructions are in the [Helium readme](https://github.com/retaildevcrews/helium)
  - Please complete the setup steps and then continue below

### Validate az CLI works

> In GitHub Codespaces, open a terminal by pressing ctl + `

```bash

# make sure you are logged into Azure
az account show

# if not, log in
az login

```

### Verify Key Vault Access

```bash

# verify you have access to Key Vault
az keyvault secret show --name CosmosDatabase --vault-name $He_Name

```

### Using GitHub Codespaces

- Open `launch.json` in the `.vscode` directory
- Replace `{your key vault name}` with the name of your key vault
  - the file saves automatically
- Press F5
- Wait for `Application started. Press Ctrl+C to shut down.` in the Debug Console
- Skip to the testing step below

### Using bash shell

> This will work from a terminal in GitHub Codespaces as well

```bash

# run the application
# He_Name was set during setup and is your Key Vault name
dotnet run -p src/app/helium.csproj -- --auth-type CLI --keyvault-name $He_Name

```

wait for `Application started. Press Ctrl+C to shut down.`

### Testing the application

Open a new bash shell

> GitHub Codespaces allows you to open multiple shells by clicking on the `Split Terminal` icon

```bash

# test the application

# test using httpie (installed automatically in Codespaces)
http localhost:4120/version

# test using curl
curl localhost:4120/version

```

Stop helium by typing Ctrl-C or the stop button if run via F5

### Deep Testing

We use [Web Validate](https://github.com/microsoft/webvalidate) to run deep verification tests on the Web API

```bash

# install Web Validate as a dotnet global tool
# this is automatically installed in CodeSpaces
dotnet tool install -g webvalidate

# make sure you are in the root of the repository

# run the validation tests
# validation tests are located in the TestFiles directory

pushd TestFiles
webv -s localhost:4120 -f baseline.json

# there may be a validation error on the /healthz/ietf test
#   json: status: warn : Expected: pass
# the "warn" status indicates a slower than normal response time
# and will occasionally occur due to network latency

# bad.json tests error conditions that return 4xx codes

# benchmark.json is a 300 request test that covers the entire API

popd

```

## Build the release container using Docker

> A release build requires MI to connect to Key Vault.

```bash

# make sure you are in the root of the repo
# build the image
docker build . -t helium-csharp

# run docker tag and docker push to push to your repo

```

## CI-CD

> Make sure to fork the repo before experimenting with CI-CD

This repo uses [GitHub Actions](.github/workflows/dockerCI.yaml) for Continuous Integration.

- CI supports pushing to Azure Container Registry or DockerHub
- The action is setup to execute on a PR or commit to ```master```
  - The action does not run on commits to branches other than ```master```
- The action always publishes an image with the ```:beta``` tag
- If you tag the repo with a version i.e. ```v1.0.8``` the action will also
  - Tag the image with ```:1.0.8```
  - Tag the image with ```:stable```
  - Note that the ```v``` is case sensitive (lower case)
- Once the `secrets` below are set, create a new branch, make a change to a file (md file changes are ignored), commit and push your change, create a PR into your local master
- Check the `Actions` tab on the GitHub repo main page

CD is supported via webhooks in Azure App Services connected to the ACR or DockerHub repository.

### CI to Azure Container Registry

In order to push to ACR, you set the following `secrets` in your GitHub repo:

- Azure Login Information
  - TENANT
  - SERVICE_PRINCIPAL
  - SERVICE_PRINCIPAL_SECRET

- ACR Information
  - ACR_REG
  - ACR_REPO
  - ACR_IMAGE

### CI to DockerHub

In order to push to DockerHub, you must set the following `secrets` in your GitHub repo:

- DOCKER_REPO
- DOCKER_USER
- DOCKER_PAT
  - Personal Access Token (recommended) or password

## Contributing

This project welcomes contributions and suggestions. Most contributions require you to agree to a
Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us
the rights to use your contribution. For details, visit [Microsoft Contributor License Agreement](https://cla.opensource.microsoft.com).

When you submit a pull request, a CLA bot will automatically determine whether you need to provide
a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions
provided by the bot. You will only need to do this once across all repos using our CLA.

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.


CHANGELOG.md
# Change Log

## 1.6.0

- Oct 2021
- updated dev container / Codespaces build

## v1.6.0

- March 2020
- added authentication type environment varible and command line requirements
  - supports MSI, Azure CLI or Visual Studio credentials
- converted Swagger to not auto-generate swagger.json for fidelity

## v1.0.0

- Dec 2019
- converted /healthz to use dotnet HealthCheck
- added IETF draft render to /healthz/ietf

## Dec 2019

- updated md docs

## Nov 2019

- Created a separate repo for the application
- Upgraded to dotnet core 3.0

## Oct 2019

- added retry for AKS pod identity (it takes ~ 30 seconds to spin up the first time in the cluster)
- added /version endpoint
- added additional movie APIs
  - /api/movies?genre=action
  - /api/movies?year=1999
  - /api/movies?rating=8.5 (this returns >= 8.5)
  - /api/movies?actorid=nm0000206
  - /api/movies?toprated=true (returns the 10 top rated movies sorted by rating desc)
  - /api/featured/movie (returns a random movie from the featured movies)

## Oct 2019

- Enhanced exception logging to include CosmosDB information where available
- Simplified dockerfile

## Sept 2019

- Added logic to support CosmosDB key rotation
- Added middleware to handle /robots*.txt requests
- Added integration test (to dotnet.json) for robots*.txt
- added dockerfile to build dev environment
- added optional proxy variables in dockerfile
- added .dockerignore
- simplified dockerfile
- renamed appsettings.json and made it required
- removed unused assemblies
- removed and sorted usings
- updated to new IMDb extract
- added totalScore to Movie for upcoming "voting version"
- removed "profession" from Movies.Roles for clarity
- renamed /key to /partitionKey for clarity
- added call to dal.Healthz() on DAL create and reload
  - if CreateDal() fails, the app will exit(-1) - "fail early"
  - if ReloadDal() fails, an error is logged and the old DAL is used until the next check
- renamed class IDal to DAL for clarity
- added asserts to check /partitionKey in unit tests

## Sept 5, 2019

- Initial commit


src/e2e-tests/RunApp.cs
using Helium.Controllers;
using Helium.Model;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using Moq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using Xunit;

namespace UnitTests
{
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1001:Types that own disposable fields should be disposable")]
    public class AppTest
    {
        HttpResponseMessage resp;
        string res;

        [Fact]
        public async Task RunApp()
        {
            string[] args = new string[] { "--kvname", "bluebell-kv", "--authtype", "CLI"  };

            await Helium.App.Main(null);
            await Helium.App.Main(new string[] { "--help" });
            await Helium.App.Main(new string[] { "--kvname", " " });

            Task<int> t = Helium.App.Main(args);

            await Task.Delay(10000);


            using HttpClient client = new HttpClient { BaseAddress = new System.Uri("http://localhost:4120") };

            try
            {
                res = await client.GetStringAsync("/healthz/dotnet");
                res = await client.GetStringAsync("/index.html");
                res = await client.GetStringAsync("/swagger.json");
                res = await client.GetStringAsync("/version");
                res = await client.GetStringAsync("/robots.txt");
                res = await client.GetStringAsync("/api/genres");

                res = await client.GetStringAsync("/api/actors?pageNumber=2&pageSize=10");
                res = await client.GetStringAsync("/api/actors?q=nicole&pageNumber=1&pageSize=10");
                res = await client.GetStringAsync("/api/actors/nm0000173");
                res = await client.GetStringAsync("/api/actors/nm0000031");

                res = await client.GetStringAsync("/healthz");

                res = await client.GetStringAsync("/api/movies?pageNumber=2&pageSize=10");
                res = await client.GetStringAsync("/api/movies?q=ring&pageNumber=1&pageSize=10");
                res = await client.GetStringAsync("/api/featured/movie");
                res = await client.GetStringAsync("/api/movies/tt0133093");
                res = await client.GetStringAsync("/api/movies?actorId=nm0000206");
                res = await client.GetStringAsync("/api/movies?rating=8.5");
                res = await client.GetStringAsync("/api/movies?year=1999");
                res = await client.GetStringAsync("/api/movies?genre=action");
                res = await client.GetStringAsync("/healthz/ietf");

                resp = await client.GetAsync("/api/movies?pageSize=-1");
                resp = await client.GetAsync("/api/movies?pageSize=1001");
                resp = await client.GetAsync("/api/movies?pageSize=10.1");
                resp = await client.GetAsync("/api/movies?pageSize=foo");
                resp = await client.GetAsync("/api/movies?pageNumber=0");
                resp = await client.GetAsync("/api/movies?pageNumber=-1");
                resp = await client.GetAsync("/api/movies?pageNumber=10.1");
                resp = await client.GetAsync("/api/movies?pageNumber=10001");
                resp = await client.GetAsync("/api/movies?pageNumber=foo");
                resp = await client.GetAsync("/api/movies?year=foo");
                resp = await client.GetAsync("/api/movies?year=-1");
                resp = await client.GetAsync("/api/movies?year=0");
                resp = await client.GetAsync("/api/movies?year=1");
                resp = await client.GetAsync("/api/movies?year=1873");
                resp = await client.GetAsync("/api/movies?year=2026");
                resp = await client.GetAsync("/api/movies?year=2020.1");
                resp = await client.GetAsync("/api/movies?year=2025");
                resp = await client.GetAsync("/api/movies?year=1874");
                resp = await client.GetAsync("/api/movies?rating=foo");
                resp = await client.GetAsync("/api/movies?rating=-1");
                resp = await client.GetAsync("/api/movies?rating=10.1");
                resp = await client.GetAsync("/api/movies?genre=ab");
                resp = await client.GetAsync("/api/movies?genre=123456789012345678901");
                resp = await client.GetAsync("/api/movies?actorId=nm123");
                resp = await client.GetAsync("/api/movies?actorId=ab12345");
                resp = await client.GetAsync("/api/movies?actorId=tt12345");
                resp = await client.GetAsync("/api/movies?actorId=NM12345");
                resp = await client.GetAsync("/api/movies?actorId=nm12345");
                resp = await client.GetAsync("/api/actors/ab12345");
                resp = await client.GetAsync("/api/actors/tt12345");
                resp = await client.GetAsync("/api/actors/NM12345");
                resp = await client.GetAsync("/api/actors/nm123");
                resp = await client.GetAsync("/api/actors/nmabcde");
                resp = await client.GetAsync("/api/actors/nm00000");
                resp = await client.GetAsync("/api/actors/nm12345");
                resp = await client.GetAsync("/api/actors/nm0000173/foo");
                resp = await client.GetAsync("/api/movies/ab12345");
                resp = await client.GetAsync("/api/movies/nm12345");
                resp = await client.GetAsync("/api/movies/TT12345");
                resp = await client.GetAsync("/api/movies/tt123");
                resp = await client.GetAsync("/api/movies/ttabcde");
                resp = await client.GetAsync("/api/movies/tt00000");
                resp = await client.GetAsync("/api/movies/tt12345");
                resp = await client.GetAsync("/api/movies/tt0133093/foo");
            }

            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }

            Helium.App.Stop();

            await Task.Delay(500);

        }

    }
}
